lib/sycamore/sycamore/document.py

```diff
@@ -0,0 +1,390 @@
+from abc import abstractmethod, ABC
+from enum import Enum
+
+
+class Category(Enum):
+    Caption = 1
+    Footnote = 2
+    Formula = 3
+    ListItem = 4
+    PageFooter = 5
+    PageHeader = 6
+    Picture = 7
+    SectionHeader = 8
+    Table = 9
+    Text = 10
+    Title = 11
+    Section = 101
+    Group = 102
+
+
+class Node:
+    def __init__(self, node_id, properties):
+        self._node_id = node_id
+        self._properties = properties
+
+    def node_id(self):
+        return self._node_id
+
+    @abstractmethod
+    def category(self):
+        pass
+
+    def properties(self):
+        return self._properties
+
+    @abstractmethod
+    def accept(self, visitor):
+        pass
+
+
+class Leaf(Node):
+    def __init__(self, node_id, box, content, properties):
+        super().__init__(node_id, properties)
+        self._box = box
+        self._content = content
+
+    def box(self):
+        return self._box
+
+    def content(self):
+        return self._content
+
+
+class Internal(Node):
+    def __init__(self, node_id, children, properties):
+        super().__init__(node_id, properties)
+        self._children = children
+
+    def children(self) -> list[Node]:
+        return self._children
+
+
+class Caption(Leaf):
+    def __init__(self, node_id, box, content, properties):
+        super().__init__(node_id, box, content, properties)
+
+    def category(self):
+        return Category.Caption
+
+    def accept(self, visitor):
+        return visitor.visit_caption(self)
```

Comment:
I would recommend moving the switch logic into the visitor class. 
In particular, I would have:
```
class Caption(Leaf):
   def __init__(self, node_id, box, content, properties):
        super().__init__(node_id, box, content, properties, category=Category.Caption)
        
class Leaf(...):
   def category(self):
        return self.category

   def accept(self, visitor):
       return visitor.visit(self, self.category) # optional to just have self since the category is bound

class Visitor():
   def visit(self, node, category):
        if category == Category.Caption:
            self.visit_caption(node)
       elif ...
```

That way someone can override just the visit function if they want to handle the grouping differently rather than having to implement a whole bunch of functions.

You may also want to implement more flexible visiting. For plan nodes we needed pre-traversal, post-traversal, in-place mutation, and single call for the entire tree.
See plan_nodes.py:NodeTraverse.  My guess is that if we're going to do mutations of the structure via traversal, we'll end up needing the same thing.